randovania/resolver/bootstrap.py

```diff
@@ -250,29 +251,57 @@ def all_preplaced_pickup_locations(
 
         return locations
 
-    def pre_place_pickups(
+    def pre_place_pickups_weighted(
         self,
         rng: Random,
-        locations: list[PickupNode],
+        locations: dict[PickupNode, float],
         pool_results: PoolResults,
         item_category: str,
         game: RandovaniaGame,
     ) -> None:
         pre_placed_indices = list(pool_results.assignment.keys())
-        reduced_locations = [loc for loc in locations if loc.pickup_index not in pre_placed_indices]
+        reduced_locations = [loc for loc in locations.keys() if loc.pickup_index not in pre_placed_indices]
 
-        rng.shuffle(reduced_locations)
+        # working_locations is a copy of locations with only the keys in reduced locations
+        working_locations = {}
+        for loc in reduced_locations:
+            working_locations[loc] = locations[loc]
+
+        # weighted locations is a list filled by selecting weighted elements from working locations
+        weighted_locations = []
+        for _ in reduced_locations:
+            loc = random_lib.select_element_with_weight_and_uniform_fallback(rng, working_locations)
+            weighted_locations.append(loc)
+            working_locations.pop(loc)
```

Comment:
...and this block would be 
```python
        while bool(reduced_locations):
            loc = random_lib.select_element_with_weight_and_uniform_fallback(rng, reduced_locations)
            weighted_locations.append(loc)
            reduced_locations.pop(loc)
```
because you iterate over it until it's empty